### ISSUE TITLE
Additional updates for layers

### ISSUE THREAD

**Issue by diogo149** [benanne/Lasagne#136]
While working on https://github.com/benanne/Lasagne/issues/133, I realized that there isn't a way for a layer to update it's state outside of the normal `updates.py` updates. Specifically, weight decay for learnable activation units and rolling averages for batch normalization.

One proposed solution would be to have some additional `get_additional_updates` method, like `get_params`, that gets merged into the updates. Implementation should be pretty clean/straightforward if one is careful to merge the updates together (since each var can only be updated once, and you don't want to double add the var).

**Comment by benanne** [comment]
I'm not entirely sure I understand what problems this solves. This is an interface change, and more importantly it makes the interface bigger, so I always try to be very critical of such modifications. The smaller we can keep it, the better.

Could you explain a use case for this more concretely and maybe list some others as well? I still need to be convinced that this is worth the additional complexity :p

I sort of understand why it would be useful for rolling averages, but I don't understand why weight decay for learnable activation units requires this.

**Comment by diogo149** [comment]
Perhaps that solution was a bad one, so I'll focus on the problem instead. (: I think the problem to solve is non-homogeneous updates in a somewhat encapsulated way. Some use cases would be (a) for LA units, the authors strongly recommend L2 loss, (b) same thing for a L2 hingle output layer, and (c) rolling averages for batch normalization. My understanding of the current way of solving this would be for a user to access the fields of a layer and manually update the objective or updates, which makes anything that relies on that functionality significantly less simple to use.

I'm not the most familiar about where Lasagne wants to stand in the spectrum of user friendly and extensible, or what kind of use cases you want to solve, so whether or not that's a problem worth solving is up to you (:

**Comment by benanne** [comment]
We want to make the common use cases easy and the less common ones possible - but if we can make some less common ones easier without it affecting the common ones, we should do that. In this, it's important not to forget that "extending the library" is also considered an extremely common use case for us, and every little bit of "interface" adds some complexity that makes understanding and extending the library harder. This is why I'm so conservative in this regard.

I'm not entirely convinced that the use cases you propose require an extension of the layer interface. For some it seems like this would better be handled externally (especially (a) and (b)), but we can definitely provide convenience functions for this. It is also possible for a `Layer` class to define methods that are not part of the common `Layer` interface. When we establish that other types of layers could benefit from this interface as well, we can always generalize it later.

To me the computation of rolling averages in batch normalization is the only compelling use case so far for introducing some kind of interface for having layers generate updates by themselves (separately from the optimization algorithm). So perhaps for now it makes sense to define something like `get_additional_updates` only on a `BatchNormalizationLayer`, or something. I don't know, this seems like something we'd have to think over for a while, so we don't make any rash decisions that we'll regret later.

**Comment by diogo149** [comment]
What kind of interface were you thinking for the convenience functions for (a) and (b)? When I imagine an ideally decoupled API, I imagine being able to change anything in only one place seamlessly (eg. learnable activation unit for ReLU), and having an additional function sometimes seems non-ideal as a user. Definitely agree with being conservative on interface changes, engineering is all about tradeoffs after all. (:

I also was thinking of having layers be responsible for updating themselves. An initial thought would be a `get_all_updates`, similar to `get_all_params`, which recursively gets the updates, and layers could be passed in an update function to use (eg. `updates.momentum`) either explicitly or recursively from a different layer.

**Comment by benanne** [comment]
> I also was thinking of having layers be responsible for updating themselves. An initial thought would be a `get_all_updates`, similar to `get_all_params`, which recursively gets the updates, and layers could be passed in an update function to use (eg. `updates.momentum`) either explicitly or recursively from a different layer.

What would be the benefit of that? To me it sounds like it would be a lot more complex and somewhat counterintuitive (networks don't learn by themselves, an optimization algorithm 'teaches' them and this is external to the network). If we decide to have layers generate updates themselves, I think this should be restricted to updates that are not part of the optimization algorithm (like the rolling averages in batch normalization for example).

**Comment by diogo149** [comment]
I think it depends on where you draw the lines of separation. The advantage of something like that would be encapsulating the learning of a composable unit (encapsulating a technique, as opposed to a set of weights). The view that layers are just bunch of weights (or parameters to backprop through) and stateless ops might be a pretty big limit of flexibility. An example of the problems of this view are also in batch normalization, because a "param" (my understanding at least) is currently overloaded in referring to both a parameter and a parameter to backprop through. If batch normalization mean/variance counts as a param, then SGD update code won't work with it. If it doesn't count as a param, users relying on `get_all_params` for saving model state will have their code break (because the mean/variance won't be saved).

On the other hand, batch normalization is just one example, so fighting hard for interface changes might not be worth it. (: Thinking about the distinction between a backprop-able param and a normal param seems valuable though.

**Comment by diogo149** [comment]
Alternative serialization might just solve the `get_all_params` thing,
